File: src/main/java/org/example/dp1d/HouseRobber.java
package org.example.dp1d;

public class HouseRobber {
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        int rob1=0, rob2=0, temp;
        for (int n: nums) {
            temp = Math.max(n + rob1, rob2);
            rob1 = rob2;
            rob2 = temp;
        }
        return rob2;
    }

    public static void main(String[] args) {
        int[] nums = {1,2,3,1};
        System.out.println(new HouseRobber().rob(nums));
    }
}

File: src/main/java/org/example/dp1d/HouseRobber.py
from typing import List


class Solution:
    # 93%
    def rob(self, nums: List[int]) -> int:
        rob1, rob2 = 0, 0

        # [rob1, rob2, n, n+1, ...]
        for n in nums:
            temp = max(rob1 + n, rob2)
            rob1 = rob2
            rob2 = temp
        return rob2

    # 90%
    def rob2(self, nums: List[int]) -> int:
        if len(nums) == 2: return max(nums)

        dp = [0] * len(nums)
        dp[0] = nums[0]
        dp[1] = max(nums[0], nums[1])

        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 1], dp[i - 2] + nums[i])
        return dp[-1]


obj = Solution()
nums = [2,7,9,3,1]
print(obj.rob2(nums))
